### PR TITLE
A real pkg/cp.

### PR DESCRIPTION
**Pull request by hugelgupf**
I really wanted a cp.CopyTree function, esp for #1296. So I went and made one. In the process, I went a bit overboard and re-arranged cmds/core/cp and pkg/cp.

Also fixes one small issue:
- symlink copying didn't imitate GNU cp. I honestly don't like either solution, so I'm just going with the GNU one. GNU copies the symlink target name, regardless of what it is or what it is relative to. 

Also removes the parallel copying. If that is to make a comeback, it needs an implementation with targeted unit tests and benchmarks in `pkg/cp`.

### CODE REVIEW

File: pkg/cp/cmp/cmp.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+func stat(o cp.Options, path string) (os.FileInfo, error) {
+	if o.NoFollowSymlinks {
+		return os.Lstat(path)
+	}
+	return os.Stat(path)
+}
+
+// IsEqualTree compare the content between of src and dst paths
```
Comment (rminnich) [inline pkg/cp/cmp/cmp.go]: between of?

File: pkg/cp/cmp/cmp.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	case srcInfo.Mode().IsDir():
+		srcEntries, err := readDirNames(src)
+		if err != nil {
+			return err
+		}
+		dstEntries, err := readDirNames(dst)
+		if err != nil {
+			return err
+		}
+		if !reflect.DeepEqual(srcEntries, dstEntries) {
```
Comment (rminnich) [inline pkg/cp/cmp/cmp.go]: is this assuming they names are in the directory in the same order? If so that seems unsafe to me.

File: pkg/cp/cp.go [hunk 3]
```diff
@@ -2,27 +2,148 @@
...
+func copyFile(src, dst string, srcInfo os.FileInfo) error {
+	m := srcInfo.Mode()
+	switch {
+	case m.IsDir():
+		return os.Mkdir(dst, srcInfo.Mode().Perm())
+
+	case m.IsRegular():
+		return copyRegularFile(src, dst, srcInfo)
+
+	case m&os.ModeSymlink == os.ModeSymlink:
```
Comment (rminnich) [inline pkg/cp/cp.go]: did there need to be a test in here for following symlinks? I got a bit lost.

File: pkg/cp/cmp/cmp.go (hunk 2 above)
Comment (hugelgupf) [reply pkg/cp/cmp/cmp.go]: ioutil.ReadDir guarantees they are sorted.

File: pkg/cp/cp.go (hunk 3 above)
Comment (hugelgupf) [reply pkg/cp/cp.go]: No -- because if you used Stat, this would not show up as a symlink unless it couldn't be followed.

**Comment by hugelgupf** [comment]
PTAL

**Pull request merged**